### ISSUE TITLE
Local vs global error handling

### ISSUE THREAD

**Issue by nimmolo** [DocWatson/wufoo-ajax-helper#1]
Hi Doc,
I'm using an adaptation of your code (thanks!) and i'm having trouble figuring out how to handle local errors. 

With the script, it seems only an actual HTML error communicating with Wufoo triggers the error case. Form field errors do not trigger the error case, even though they log a value of "0" for "success".

When a form field has a validation error, the wufoo js executes the conditional for a success, simply logging the response (which includes a FeldError array) and displaying my success message.

Sorry if this question is too noob... I just need help writing the conditional of a local validation error, so i can loop through the field labels and replace them with error messages in that case.

Here's the event log, triggered by an invalid email address

```
Object
EntryLink: null
ErrorText: "Errors have been <b>highlighted</b> below."
FieldErrors: Array[1]
0: Object
ErrorText: "Please enter a valid email address."
ID: "Field6"
__proto__: Object
length: 1
__proto__: Array[0]
Success: 0
__proto__: Object
```

And here's the script i'm using, simplified

```
jQuery('.wufoo-form').on('submit', function(e){
  e.preventDefault();
  var formType = jQuery(this).data('form-type');

  if (formType === undefined) {
    console.log('ERROR: Form type not defined. Include a data-form-type attribute on your form tag');
    return;
  }

  var data = {
    action    : 'wufoo_post',
    fields    : jQuery(this).serialize(),
    form_type : formType
  };

  jQuery.ajax({
    type     : 'POST',
    url      : '/wp-admin/admin-ajax.php',
    data     : data,
    dataType : 'json',
    success  : function(response) {
      console.log(response);
      // show success message
    },
    error : function(jqXHR, textStatus, errorThrown) {
      console.error(errorThrown);
      // show general error message
    }
  });
});
```

**Comment by DocWatson** [comment]
Hi there,

I was actually adapting this plugin for another project and noticed that this part wasn't finished. I'd need to do a bit of coding and testing to get it right, but you'd want to do something like:

```
if (response.success !== 1) {
   //we have an error, so loop through the errors
  jQuery(response.FieldErrors).each(function(){
    //target the erroneous field by ID and add an error class
    jQuery(this.ID).addClass("error");
    //display error message after erroneous field
    jQuery(this.ID).insertAfter(this.ErrorText);
  });
}
```

The code is a little sloppy because I'm writing from memory (stuck very late at my day job), but that's the gist of what you need to do.

Alternatively, you could use a separate library to do your validation and not rely on the Wufoo response (though I'd recommend doing both). If you're looking for a suggestion, I'm partial to [parsley.js](https://github.com/guillaumepotier/Parsley.js).

Also, make sure your inputs are named the way wufoo is expecting them (e.g. your field for email address should have an attribute `name="Field6"). You probably know that, but that's something I forget to do from time to time :).

If you get it working, feel free to send me a pull request! I'll try to update this to be a little more sane in the next day or two. Sorry for the trouble!

**Comment by nimmolo** [comment]
Many thanks, that pretty much cracked it. 
I just made a few tweaks to get it working. Note capitalization in (response.Success) and a slightly different syntax for the each() loop.

```
      jQuery.ajax({
        type     : 'POST',
        url      : '/wp-admin/admin-ajax.php',
        data     : data,
        dataType : 'json',
        success  : function(response) {
          console.log(response);
          if (response.Success !== 1) {
            // we have a field error, so loop through the errors
            var errors = response.FieldErrors;
            jQuery.each( errors, function(){
              // in case of multiple errors let's log them with their fieldID's
              console.log(this.ID + ": " + this.ErrorText);
              // target the erroneous fields by ID and add an error class
              jQuery('input#' + this.ID ).addClass("error");
              // replace label with error message
              jQuery('label[for=' + this.ID + ']').addClass("error").html(this.ErrorText);
            });
          } else {
              // we're good
              console.log("Post successful");
          }
        },
        error : function(jqXHR, textStatus, errorThrown) {
          // HTML error in communicating with Wufoo
          console.error(errorThrown);
        }
      });
```

This could be simpler, it reflects the fact that in my form layout, i print the error messages within the input label.  

_edit: moved "new issue" to new issue thread_

**Comment by nimmolo** [comment]
OK pull request is in. 

I'm a total newb with GitHub so if i've done something wrong in the pull request, please let me know.

**Comment by nimmolo** [comment]
Yikes- just realizing this didnt send when i wrote it two days ago. Pull req should be in by now! :)

---

Wicked! 

Thanks - i will try as soon as i get back. And will submit pull req if i get it working. 

no trouble at all and tx for responding

**Comment by DocWatson** [comment]
Sorry for the slow response - I have merged it in! Thanks for your patience. Have you solved your issues, or do you still have an outstanding question? I should have more free time this week (finally!) to work on my pet projects.

**Comment by nimmolo** [comment]
Great - all good. Working well. 

The "New issue" above would be nice to tackle, for longer forms with multiple validations, but i should open a new issue. In any case it's not essential for me for this form. It seems like there's probably already a "best practice" for this issue on ajax forms, i just don't know it.
